File: app/src/main/java/pl/srw/billcalculator/util/BillSelection.java
package pl.srw.billcalculator.util;

import android.util.SparseArray;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;

import javax.inject.Inject;
import javax.inject.Singleton;

import hugo.weaving.DebugLog;
import pl.srw.billcalculator.db.Bill;

@Singleton
public class BillSelection {

    private SparseArray<Bill> selectedItems;

    @Inject
    public BillSelection() {
        selectedItems = new SparseArray<>();
    }

    @DebugLog
    public void select(final int position, final Bill o) {
        selectedItems.put(position, o);
    }

    @DebugLog
    public void deselect(final int position) {
        selectedItems.remove(position);
    }

    public boolean isSelected(final int position) {
        return selectedItems.indexOfKey(position) >= 0;
    }

    public boolean isAnySelected() {
        return selectedItems.size() > 0;
    }

    public void deselectAll() {
        selectedItems.clear();
    }

    public Collection<Bill> getItems() {
        final ArrayList<Bill> values = new ArrayList<>(selectedItems.size());
        for (int i = 0; i < selectedItems.size(); i++) {
            values.add(selectedItems.valueAt(i));
        }
        return values;
    }

    public int[] getPositionsReverseOrder() {
        int[] keys = new int[selectedItems.size()];
        for (int i = 0; i < selectedItems.size(); i++) {
            keys[i] = selectedItems.keyAt(i);
        }
        Arrays.sort(keys);
        return Sorter.revert(keys);
    }
}

File: app/src/main/java/pl/srw/billcalculator/util/BillSelection.kt
package pl.srw.billcalculator.util

import pl.srw.billcalculator.db.Bill
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class BillSelection @Inject constructor() {

    private val selectedItems = mutableMapOf<Int, Bill>()

    fun select(position: Int, o: Bill) {
        selectedItems[position] = o
    }

    fun deselect(position: Int) {
        selectedItems.remove(position)
    }

    fun deselectAll() {
        selectedItems.clear()
    }

    fun isSelected(position: Int) = selectedItems.containsKey(position)

    fun isAnySelected() = selectedItems.isNotEmpty()

    fun getItems(): Collection<Bill> = selectedItems.values

    fun getPositionsReverseOrder() = selectedItems.keys.sortedDescending().toIntArray()

    fun onInsert(atPosition: Int) {
        val updated = mutableMapOf<Int, Bill>()
        val iterator = selectedItems.iterator()
        while (iterator.hasNext()) {
            val (position, bill) = iterator.next()
            if (position >= atPosition) {
                iterator.remove()
                updated[position + 1] = bill
            }
        }
        selectedItems.putAll(updated)
    }
}
